// *****************************************************************************
//
// This file is part of a MASA library or program.
// Refer to the included end-user license agreement for restrictions.
//
// Copyright (c) 2010 MASA Group
//
// *****************************************************************************

package MsgsClientToSim;

import "Common.proto"; 

message MsgControlDatetimeChange {
   required Common.MsgDateTime date_time = 1;
}

message MsgControlCheckPointSaveNow {
   optional string name = 1;
}

message MsgFragOrder {
   required Common.Tasker               tasker     = 1;
   required Common.FragOrderType        frag_order = 2;
   optional Common.MsgMissionParameters parameters = 3;
}

message MsgSetAutomatMode {
   required Common.AutomatId       automate = 1;
   required Common.EnumAutomatMode mode     = 2;
}

// Magic Action
message MsgUnitMagicAction {
   required Common.Tasker               tasker     = 1;
   required Common.UnitMagicActionType  type       = 2;
   required Common.MsgMissionParameters parameters = 3;
}

message MsgKnowledgeMagicAction {
   required Common.KnowledgeGroupId         knowledge_group = 1;
   required Common.KnowledgeMagicActionType type            = 2;
   required Common.MsgMissionParameters     parameters      = 3;
}

message MsgObjectMagicAction {
   enum Attribute {
        construction          = 0;
        obstacle              = 1;
        valorisation          = 2;
        activity_time         = 3;
        bypass                = 4;
        logistic              = 5;
        nbc                   = 6;
        crossing_site         = 7;
        supply_route          = 8;
        toxic_cloud           = 9;
        fire                  = 10;
        medical_treatment     = 11;
        interaction_height    = 12;
        stock                 = 13;
        nbc_agent             = 14;
        effect_delay          = 15;
        flood                 = 16;
        burn_surface          = 17;
        structural_state      = 18;
        infrastructure        = 19;
        usages                = 20;
        lodging               = 21;
        trafficability        = 22;
        underground           = 23;
   }
   
   required Common.ObjectId              object     = 1;
   required Common.ObjectMagicActionType type       = 2;
   required Common.MsgMissionParameters  parameters = 3;
}

message MsgMagicAction {
   
   required Common.MagicActionType      type       = 1;
   required Common.MsgMissionParameters parameters = 2;
}

message MsgControlCheckPointSetFrequency {
   required int32 frequency = 1;
}

message MsgControlCheckPointDeleteRequest {
   required string checkpoint = 1;
}

message MsgControlExportRequest {
   required string directory_name = 1;
}

message MsgControlCheckPointListRequest {}


message MsgControlEnableVisionCones {
   required bool vision_cones = 1;
}

message MsgOrderStream {
    required string serialized_order = 1;
}

message MsgClientToSim {
    message Content {
       optional Common.MsgControlStop                  control_stop                         = 1;
       optional Common.MsgControlPause                 control_pause                        = 2;
       optional Common.MsgControlResume                control_resume                       = 3;
       optional Common.MsgControlChangeTimeFactor      control_change_time_factor           = 4;
       optional MsgControlDatetimeChange               control_date_time_change             = 5;
       optional MsgControlCheckPointSaveNow            control_checkpoint_save_now          = 6;
       optional MsgControlCheckPointSetFrequency       control_checkpoint_set_frequency     = 7;
       optional MsgControlEnableVisionCones            control_toggle_vision_cones          = 8;
       optional Common.MsgUnitOrder                    unit_order                           = 9;
       optional Common.MsgAutomatOrder                 automat_order                        = 10;
       optional Common.MsgCrowdOrder                   crowd_order                          = 11;
       optional MsgFragOrder                           frag_order                           = 12;
       optional MsgSetAutomatMode                      set_automat_mode                     = 13;
       optional MsgUnitMagicAction                     unit_magic_action                    = 14;
       optional MsgObjectMagicAction                   object_magic_action                  = 15;
       optional MsgKnowledgeMagicAction                knowledge_magic_action               = 16;
       optional MsgMagicAction                         magic_action                         = 17;
       optional MsgControlCheckPointListRequest        control_checkpoint_list_request      = 18;
       optional MsgControlCheckPointDeleteRequest      control_checkpoint_delete_request    = 19;
       optional MsgControlExportRequest                control_export                       = 20;
       optional MsgOrderStream                         order_stream                         = 21;
    }
    optional int32   context = 1;
    required Content message = 2;
}
